contrib/pkg/v1migration/restore.go

```diff
@@ -0,0 +1,164 @@
+		}
+		obj, err := referencorClient(client, ref).Get(ref.Name, metav1.GetOptions{})
+		if err != nil {
+			logger.WithError(err).Error("could not get object to restore owner reference")
+			continue
+		}
+		logger = logger.WithField("referencedKind", ref.Ref.Kind).WithField("referencedName", ref.Ref.Name)
+		referencedObj, err := referenceeClient(client, ref).Get(ref.Ref.Name, metav1.GetOptions{})
+		if err != nil {
+			logger.WithError(err).Error("could not get referenced object")
+			continue
+		}
+		if ownerRefs, changed := restoreOwnerReference(obj.GetOwnerReferences(), ref, referencedObj); changed {
+			logger.Info("restoring owner reference")
+			obj.SetOwnerReferences(ownerRefs)
+			if _, err := referencorClient(client, ref).Update(obj, metav1.UpdateOptions{}); err != nil {
+				logger.WithError(err).Error("could not update object")
+			}
+		} else {
+			logger.Info("owner reference already restored")
+		}
+	}
+	return nil
+}
+
+func referencorClient(client dynamic.Interface, ref ownerRef) dynamic.ResourceInterface {
+	gvr := schema.GroupVersionResource{
+		Group:    ref.Group,
+		Version:  ref.Version,
+		Resource: ref.Resource,
+	}
+	if ns := ref.Namespace; ns != "" {
+		return client.Resource(gvr).Namespace(ns)
+	}
+	return client.Resource(gvr)
+}
+
+func referenceeClient(client dynamic.Interface, ref ownerRef) dynamic.ResourceInterface {
+	gvr := hivev1.SchemeGroupVersion.WithResource(resourceForHiveKind(ref.Ref.Kind))
+	if isNamespaceScoped(ref.Ref.Kind) {
+		return client.Resource(gvr).Namespace(ref.Namespace)
+	}
+	return client.Resource(gvr)
+}
+
+func restoreOwnerReference(ownerRefs []metav1.OwnerReference, ref ownerRef, referencedObj *unstructured.Unstructured) (newOwnerRefs []metav1.OwnerReference, changed bool) {
+	newUID := referencedObj.GetUID()
+	for i, ownerRef := range ownerRefs {
+		if ownerRef.APIVersion != ref.Ref.APIVersion ||
+			ownerRef.Kind != ref.Ref.Kind ||
+			ownerRef.Name != ref.Ref.Name {
+			continue
+		}
+		if ownerRef.UID != newUID {
+			ownerRefs[i].UID = newUID
+			ownerRefs[i].BlockOwnerDeletion = ref.Ref.BlockOwnerDeletion
+			ownerRefs[i].Controller = ref.Ref.Controller
+			changed = true
+		}
+		return ownerRefs, changed
+	}
+	ownerRef := ref.Ref
+	ownerRef.UID = newUID
+	return append(ownerRefs, ownerRef), true
+}
+
+func resourceForHiveKind(kind string) string {
+	// We can do this more safely by getting the resource using the discovery client, but we are only dealing with
+	// Hive resources, for which we know that the resource name will fit the expected pattern.
+	return strings.ToLower(kind) + "s"
+}
+
+func isNamespaceScoped(kind string) bool {
+	// We can do this more safely by getting the resource using the discovery client, but we are only dealing with
+	// Hive resources, for which we know whether the resource is namespace-scoped or cluster-scoped.
+	switch kind {
+	case "ClusterImageSet", "HiveConfig", "SelectorIdentityProvider", "SelectorSyncSet":
```

Comment:
I believe should be SelectorSyncIdentityProvider. Weirdly we have a handful of refs to this same typo already but it appears to all be in comments/godoc.

Reply:
Doh